### PR TITLE
demangler: Add an executable "demangler" file

### PR DESCRIPTION
**Pull request by honggyukim**
It'd be better to have an independent demangler binary.
The usage is as follows:

    $ demangler --help
    Usage: demangler [OPTION...] [<mangled symbol>]
    demangler -- internal simple demangler
    
          --full                 Use cxx demangler
          --no                   Do not use demangler
          --simple               Use internal simple demangler (default)
      -?, --help                 Give this help list
          --usage                Give a short usage message
      -V, --version              Print program version
    
    It uses simple demangle by default.

Signed-off-by: Honggyu Kim <honggyu.kp@gmail.com>

### CODE REVIEW

**Comment by namhyung** [comment]
I think we can reduce the build dependency [like below](https://github.com/namhyung/uftrace/files/1617688/build-reduce-dependency-of-demangler.patch.txt)?  Also I think it's better moving it to the misc/ directory.

```diff
diff --git a/Makefile b/Makefile
index b2b20a8..b9a6d37 100644
--- a/Makefile
+++ b/Makefile
@@ -99,7 +99,7 @@ TARGETS  := $(patsubst %,$(objdir)/%,$(_TARGETS))
 UFTRACE_SRCS := $(srcdir)/uftrace.c $(wildcard $(srcdir)/cmd-*.c $(srcdir)/utils/*.c)
 UFTRACE_OBJS := $(patsubst $(srcdir)/%.c,$(objdir)/%.o,$(UFTRACE_SRCS))
 
-DEMANGLER_SRCS := $(srcdir)/demangler.c $(wildcard $(srcdir)/cmd-*.c $(srcdir)/utils/*.c)
+DEMANGLER_SRCS := $(srcdir)/demangler.c $(srcdir)/utils/demangle.c $(srcdir)/utils/debug.c
 DEMANGLER_OBJS := $(patsubst $(srcdir)/%.c,$(objdir)/%.o,$(DEMANGLER_SRCS))
 
 UFTRACE_ARCH_OBJS := $(objdir)/arch/$(ARCH)/uftrace.o
@@ -132,7 +132,6 @@ CFLAGS_$(objdir)/mcount.op = -pthread
 CFLAGS_$(objdir)/cmd-record.o = -DINSTALL_LIB_PATH='"$(libdir)"'
 CFLAGS_$(objdir)/cmd-live.o = -DINSTALL_LIB_PATH='"$(libdir)"'
 LDFLAGS_$(objdir)/uftrace = -L$(objdir)/libtraceevent -ltraceevent -ldl
-LDFLAGS_$(objdir)/demangler = -L$(objdir)/libtraceevent -ltraceevent -ldl
 
 LIBMCOUNT_FAST_CFLAGS := -DDISABLE_MCOUNT_FILTER
 LIBMCOUNT_SINGLE_CFLAGS := -DSINGLE_THREAD
@@ -216,8 +215,8 @@ $(srcdir)/utils/auto-args.h: $(srcdir)/misc/prototypes.h $(srcdir)/misc/gen-auto
 $(objdir)/uftrace: $(UFTRACE_OBJS) $(UFTRACE_ARCH_OBJS) $(objdir)/libtraceevent/libtraceevent.a
        $(QUIET_LINK)$(CC) $(UFTRACE_CFLAGS) -o $@ $(UFTRACE_OBJS) $(UFTRACE_ARCH_OBJS) $(UFTRACE_LDFLAGS)
 
-$(objdir)/demangler: $(DEMANGLER_OBJS) $(UFTRACE_ARCH_OBJS) $(objdir)/libtraceevent/libtraceevent.a
-       $(QUIET_LINK)$(CC) $(DEMANGLER_CFLAGS) -o $@ $(DEMANGLER_OBJS) $(UFTRACE_ARCH_OBJS) $(DEMANGLER_LDFLAGS)
+$(objdir)/demangler: $(DEMANGLER_OBJS)
+       $(QUIET_LINK)$(CC) $(DEMANGLER_CFLAGS) -o $@ $(DEMANGLER_OBJS) $(DEMANGLER_LDFLAGS)
 
 install: all
        $(Q)$(INSTALL) -d -m 755 $(DESTDIR)$(bindir)
```

**Comment by honggyukim** [comment]
It's a bit simple but would it be better to add your signed-off-by in this case?
I've updated including your change anyway. Thanks a lot!

**Comment by namhyung** [comment]
No I don't need it for this :)

**Comment by honggyukim** [comment]
Just asked as an example. Thanks for your help!

**Comment by honggyukim** [comment]
FYI, I didn't add this for installation.

File: Makefile [hunk 1]
```diff
@@ -91,11 +93,15 @@ LIBMCOUNT_TARGETS += libmcount/libmcount-single.so libmcount/libmcount-fast-sing
...
 _TARGETS := uftrace libtraceevent/libtraceevent.a
 _TARGETS += $(LIBMCOUNT_TARGETS) libmcount/libmcount-nop.so
+_TARGETS += misc/demangler
 TARGETS  := $(patsubst %,$(objdir)/%,$(_TARGETS))
 
 UFTRACE_SRCS := $(srcdir)/uftrace.c $(wildcard $(srcdir)/cmd-*.c $(srcdir)/utils/*.c)
 UFTRACE_OBJS := $(patsubst $(srcdir)/%.c,$(objdir)/%.o,$(UFTRACE_SRCS))
 
+DEMANGLER_SRCS := $(srcdir)/misc/demangler.c $(srcdir)/utils/demangle.c $(srcdir)/utils/debug.c
+DEMANGLER_OBJS := $(patsubst $(srcdir)/misc/%.c,$(objdir)/misc/%.o,$(DEMANGLER_SRCS))
```
Comment (namhyung) [inline Makefile]: No, it won't match `$(srcdir)/utils/*` anymore.  You need to remove the `/misc` part IMO.

**Comment by namhyung** [comment]
In the future, it'd be better to make it resemble to `c++filt` so that it can demangle all symbols in the stdin (or a given file).

File: Makefile (hunk 1 above)
Comment (honggyukim) [reply Makefile]: Like this way?
```diff
diff --git a/Makefile b/Makefile
index 8e63ff6..805bfed 100644
--- a/Makefile
+++ b/Makefile
@@ -100,7 +100,7 @@ UFTRACE_SRCS := $(srcdir)/uftrace.c $(wildcard $(srcdir)/cmd-*.c $(srcdir)/utils
 UFTRACE_OBJS := $(patsubst $(srcdir)/%.c,$(objdir)/%.o,$(UFTRACE_SRCS))
 
 DEMANGLER_SRCS := $(srcdir)/misc/demangler.c $(srcdir)/utils/demangle.c $(srcdir)/utils/debug.c
-DEMANGLER_OBJS := $(patsubst $(srcdir)/misc/%.c,$(objdir)/misc/%.o,$(DEMANGLER_SRCS))
+DEMANGLER_OBJS := $(patsubst $(srcdir)/%.c,$(objdir)/%.o,$(DEMANGLER_SRCS))
 
 UFTRACE_ARCH_OBJS := $(objdir)/arch/$(ARCH)/uftrace.o 
```

**Comment by honggyukim** [comment]
> In the future, it'd be better to make it resemble to c++filt so that it can demangle all symbols in the stdin (or a given file).

I think it's a good idea. And if so, what about changing its name from `demangler` to `c++demangler` then?

**Comment by honggyukim** [comment]
I just updated it again with your change and used the name `demangler` as is.

**Comment by Taeung** [comment]
@honggyukim , @namhyung 
Isn't needed to modify .gitignore ?
Because of below:
```
$ make
$ git status
On branch refactor-script-func
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	misc/demangler

nothing added to commit but untracked files present (use "git add" to track)
```
So:
```patch
diff --git a/.gitignore b/.gitignore
index 399b2d9..484693c 100644
--- a/.gitignore
+++ b/.gitignore
@@ -35,3 +35,4 @@ cscope.*
 *.sw[opn]
 *.patch
 *.orig
+misc/demangler

```

**Comment by honggyukim** [comment]
@Taeung You're right. Could you please add this?

**Comment by namhyung** [comment]
yes, please

**Comment by Taeung** [comment]
Yep, I'm late, I'll send PR for it.

**Pull request merged**